gossipd/gossip_store.c

```diff
@@ -135,6 +144,9 @@ bool gossip_store_read_next(struct routing_state *rstate,
 	} else if(type == WIRE_GOSSIP_STORE_NODE_ANNOUNCEMENT) {
 		fromwire_gossip_store_node_announcement(msg, msg, &gossip_msg);
 		routing_add_node_announcement(rstate, gossip_msg);
+	} else if(type == WIRE_GOSSIP_STORE_CHANNEL_DELETE) {
+		fromwire_gossip_store_channel_delete(msg, &scid);
+		tal_free(get_channel(rstate, &scid));
```

Comment:
Can `get_channel` return `NULL`?

Consider the case below:

1.  On a previous run, we added a channel, then removed it by this mechanism.  Both events are stored in the `gossip_store`.
2.  We restart and add a channel by `gossip_store`.
3.  Chaintopology catches up and notices the channel is spent.  It causes gossipd to delete the channel.
4.  We delete the channel by `gossip_store`.  At this point, `get_channel` returns `NULL`.

`tal_free` works on `NULL` though (does nothing if `NULL`).  But is the above case handled correctly?

Reply:
The situation in which we open a local channel, then spend it before it gets announced may end up in the situation you described. I'm indeed relying on the property that `tal_free(NULL)` will work in this case :wink:

For a non-local channel it is not possible that a channel is deleted without a matching announcement in the store. But I decided not to differentiate the two cases, since this is just simpler.